File: Utils/Validation/Schema/ValueSchema.cs
﻿using Newtonsoft.Json;

namespace WebCommons.Model
{
    /// <summary>
    /// Defines how a value will be validated.
    /// </summary>
    public class ValueSchema
    {
        [JsonProperty("name")]
        public string Name { get; set; } = string.Empty;

        [JsonProperty("fill", NullValueHandling = NullValueHandling.Ignore)]
        public string? Fill { get; set; }

        [JsonProperty("label", NullValueHandling = NullValueHandling.Ignore)]
        public string? Label { get; set; }

        [JsonProperty("type")]
        public string Type { get; set; } = "string";

        [JsonProperty("isRequired")]
        public bool IsRequired { get; set; } = false;

        [JsonProperty("isEnumerable")]
        public bool IsEnumerable { get; set; } = false;

        [JsonProperty("isNullable")]
        public bool IsNullable { get; set; } = false;

        [JsonProperty("min", NullValueHandling = NullValueHandling.Ignore)]
        public object? Min { get; set; } = null;

        [JsonProperty("max", NullValueHandling = NullValueHandling.Ignore)]
        public object? Max { get; set; } = null;

        [JsonProperty("regex", NullValueHandling = NullValueHandling.Ignore)]
        public string? Regex { get; set; } = null;

        [JsonProperty("values", NullValueHandling = NullValueHandling.Ignore)]
        public object[]? Values { get; set; } = null;

        [JsonProperty("default", NullValueHandling = NullValueHandling.Ignore)]
        public object? DefaultValue { get; set; } = null;

        [JsonProperty("const", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<string, object>? Constants { get; set; } = null;

        [JsonProperty("i18n", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<object, object>? I18n { get; set; } = null;

        [JsonProperty("colors", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<object, object>? Colors { get; set; } = null;

        [JsonExtensionData]
        public Dictionary<object, object> Data { get; set; } = new();
    }
}

File: Utils/Validation/Schema/ValueSchema.js
/**
 * Defines how a value will be validated.
 */
class ValueSchema {
    constructor(params, secondary) {
        params = params || {};
        secondary = secondary || {};

        this.name = params.name || secondary.name || '';
        this.fill = params.fill || secondary.fill || this.name || '';
        this.label = params.label || secondary.label || '';
        this.type = params.type || secondary.type || 'string';
        this.default = [params.default, secondary.default, null].findNotEmpty();
        this.isRequired = [params.isRequired, secondary.isRequired, false].findNotEmpty();
        this.min = [params.min, secondary.min, null].findNotEmpty();
        this.max = [params.max, secondary.max, null].findNotEmpty();
        this.regex = params.regex || secondary.regex || null;
        this.isEnumerable = [params.isEnumerable, secondary.isEnumerable, false].findNotEmpty();
        this.isNullable = [params.isNullable, secondary.isNullable, true].findNotEmpty();
        this.values = secondary.values || params.values || null;
        this.const = params.const || secondary.const || null;
        this.i18n = params.i18n || secondary.i18n || null;
    }
}